File: core/src/main/scala/diplomacy/variant/standard/rule/RetreatResolver.scala
package diplomacy.variant.standard.rule

import scala.collection.{ mutable => cmutable }

import diplomacy.board.{ DiplomacyUnit => BaseDiplomacyUnit }
import diplomacy.rule.{ InvalidOrderMessage, OrderResult, ResolvedResult }
import diplomacy.variant.standard.rule.Keywords._

class RetreatResolver extends Rule.TypeHelper {
  def apply(nextTurn: Turn => Turn)(
    board: Board, orders: Set[Order]
  ): Either[InvalidOrderMessage, ResolvedResult] = {
    val disbands = orders collect { case d@ Order.Disband(_) => d }
    val retreats = orders collect { case r@ Order.Retreat(_, _) => r }

    val results = cmutable.Map[Order, Result]()

    // Resolve disband orders
    disbands foreach { order => results(order) = Result.Success }

    // Create a map from province to retreat order
    val province2Retreats = (retreats groupBy { _.destination.province }).toMap

    // Resolve retreat orders
    province2Retreats foreach {
      case (province, retreats) if retreats.size == 1 =>
        retreats foreach { r => results(r) = Result.Success }
      case (province, retreats) if retreats.size != 1 =>
        retreats foreach { r => results(r) = Result.Failed }
    }


    // Generate a new board
    val unit2Result = results map { case (order, result) => order.unit -> (order, result) }
    val newUnits = board.units flatMap { unit =>
      unit2Result.get(unit) match {
        case Some((Order.Disband(_), Result.Success)) => None
        case Some((Order.Retreat(_, destination), Result.Success)) =>
          Option(unit.copy(location = destination))
        case Some((Order.Retreat(_, _), Result.Failed)) => None
        case _ => Option(unit)
      }
    }

    // Update occupation if needed
    val newOccupation =
      if (board.state.turn.isOccupationUpdateable) {
        board.occupation ++ (newUnits map {
          case BaseDiplomacyUnit(power, _, location) => location.province -> power
        })
      } else {
        board.occupation
      }
    val newState =
      if (board.state.turn.isBuildable) {
        board.state.copy(phase = Build)
      } else {
        board.state.copy(
          turn = nextTurn(board.state.turn),
          phase = Movement
        )
      }

    val newBoard = board.copy(
      state = newState,
      units = newUnits,
      occupation = newOccupation,
      unitStatuses = Map[DiplomacyUnit, UnitStatus](),
      provinceStatuses = Map[Province, ProvinceStatus]()
    )
    val orderResults: Set[OrderResult] = (results map {
      case (order, result) =>
        OrderResult.Executed[Power, MilitaryBranch, Order, Result](order, result)
    })(collection.breakOut)

    Right(ResolvedResult(newBoard, orderResults))
  }
}

File: src/standardRule/retreat-resolver.ts
import { Resolver } from "./resolver"
import { Result, Phase, State, MilitaryBranch, ProvinceStatus } from "./data"
import { Order, Retreat, Disband, OrderType } from "./order"
import { StandardRuleUtils } from "./standard-rule-utils"
import { Board, Unit, ResolvedResult } from "./types"
import { board } from "./../board"
import { rule } from "./../rule"
import { util } from "./../util"

declare type Province<Power> = board.Province<Power>
const { Province } = board
const { Executed } = rule
const { Success } = util

const { Movement, Build } = Phase

export class RetreatResolver<Power> implements Resolver<Power> {
  resolve (board: Board<Power>, orders: Set<Order<Power>>) {
    const disbands: Array<Disband<Power>> =
      [...orders].filter(order => order.tpe === OrderType.Disband)
    const retreats: Array<Retreat<Power>> =
      <Array<Retreat<Power>>>([...orders].filter(order => order.tpe === OrderType.Retreat))

    const retval = new Map()
    disbands.forEach(order => retval.set(order, Result.Success))

    // Create a map from province to retreat order
    const province2RetreatUnits = new Map<Province<Power>, Array<Retreat<Power>>>()
    retreats.forEach(order => {
      const elem = province2RetreatUnits.get(order.destination.province) || []
      elem.push(order)
      province2RetreatUnits.set(order.destination.province, elem)
    })

    // Resolve retreat orders
    province2RetreatUnits.forEach((orders, province) => {
      if (orders.length === 1) {
        orders.forEach(order => retval.set(order, Result.Success))
      } else {
        orders.forEach(order => retval.set(order, Result.Failed))
      }
    })

    // Generate a new board
    const unit2Result = new Map([...retval].map(elem => {
      return (<[Unit<Power>, [Order<Power>, Result]]>[elem[0].unit, [elem[0], elem[1]]])
    }))
    const newUnits = new Set()
    board.units.forEach(unit => {
      const result = unit2Result.get(unit)
      if (result) {
        const [order, r] = result
        if (order instanceof Retreat && r === Result.Success) {
          newUnits.add(new Unit(unit.militaryBranch, order.destination, unit.power))
        }
      } else {
        newUnits.add(unit)
      }
    })

    const newState = (board.state.turn.isBuildable)
      ? new State(board.state.turn, Build)
      : new State(board.state.turn.nextTurn(), Movement)

    // Update occupation if needed
    const occupationStatuses = [...board.provinceStatuses]
      .filter(elem => elem[1].occupied)
      .map(elem => {
        if (elem[1].standoff) {
          return <[Province<Power>, ProvinceStatus<Power>]>[elem[0], new ProvinceStatus(elem[1].occupied, false)]
        } else {
          return elem
        }
      })
    const newProvinceStatuses = new Map([...occupationStatuses])
    if (board.state.turn.isOccupationUpdateable) {
      newUnits.forEach(unit => {
        newProvinceStatuses.set(unit.location.province, new ProvinceStatus(unit.power, false))
      })
    }

    const newBoard = new Board(board.map, newState, newUnits, [], newProvinceStatuses)
    const orderResults = [...retval].map(elem => {
      const [order, result] = elem
      return new Executed<Power, MilitaryBranch, Result>(order, result)
    })

    const numOfCenters = ([...board.map.provinces].filter(p => p.isSupplyCenter)).length
    const numberOfSupplyCenters = StandardRuleUtils.numberOfSupplyCenters(newBoard)
    const isFinished = [...newBoard.map.powers].some(power => {
      return (numberOfSupplyCenters.get(power) || 0) > (numOfCenters / 2)
    })

    return new Success(new ResolvedResult(newBoard, orderResults, isFinished))
  }
}
